File: Examples/Advanced/Backup-Saved-Jobs/Python/backup_saved_jobs.py
#!/usr/bin/env python3
"""Back up complete netLD/ThirdEye saved-job definitions to JSON."""

from __future__ import annotations

import argparse
import json
import os
import sys
import tempfile
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterator

import requests
from dotenv import load_dotenv


FORMAT_NAME = "logicvein-netld-saved-job-backup"
FAILURE_FORMAT_NAME = "logicvein-netld-saved-job-backup-failures"
FORMAT_VERSION = 1


class ExampleError(RuntimeError):
    pass


def parse_networks(value: str) -> list[str]:
    networks = sorted({item.strip() for item in value.split(",") if item.strip()})
    if not networks:
        raise ExampleError("NETLD_NETWORKS must contain at least one managed network.")
    return networks


@dataclass(frozen=True)
class Config:
    base_url: str
    api_key: str
    networks: list[str]
    output_path: Path
    failure_path: Path
    page_size: int

    @classmethod
    def from_env(cls, env_path: Path, output_base: Path) -> "Config":
        load_dotenv(env_path, override=True)
        base_url = os.getenv("NETLD_BASE_URL", "").strip()
        api_key = os.getenv("NETLD_API_KEY", "").strip()
        if not base_url:
            raise ExampleError("Set NETLD_BASE_URL in the environment file.")
        if not api_key:
            raise ExampleError("Set NETLD_API_KEY in the environment file.")
        try:
            page_size = int(os.getenv("NETLD_JOB_PAGE_SIZE", "100"))
        except ValueError as exc:
            raise ExampleError("NETLD_JOB_PAGE_SIZE must be a positive integer.") from exc
        if page_size <= 0:
            raise ExampleError("NETLD_JOB_PAGE_SIZE must be a positive integer.")

        def destination(name: str, default: str) -> Path:
            value = Path(os.getenv(name, default))
            return value if value.is_absolute() else output_base / value

        return cls(
            base_url.rstrip("/"), api_key,
            parse_networks(os.getenv("NETLD_NETWORKS", "Default")),
            destination("NETLD_OUTPUT_FILE", "saved-jobs.json"),
            destination("NETLD_FAILURE_FILE", "saved-job-failures.json"),
            page_size,
        )


class NetLDClient:
    def __init__(self, base_url: str, api_key: str, timeout: float = 30):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"Authorization": f"Bearer {api_key}"})

    def _request(self, method: str, **kwargs: Any):
        try:
            response = self.session.request(
                method, f"{self.base_url}/rest", timeout=self.timeout,
                allow_redirects=False, **kwargs,
            )
        except requests.RequestException as exc:
            raise ExampleError(f"Could not reach {self.base_url}.") from exc
        if 300 <= response.status_code < 400:
            raise ExampleError(f"Request redirected to {response.headers.get('Location', '')}.")
        try:
            response.raise_for_status()
        except requests.RequestException as exc:
            raise ExampleError(f"Request failed with HTTP {response.status_code}.") from exc
        return response

    def login(self) -> None:
        self._request("GET")

    def call(self, method: str, **parameters: Any) -> Any:
        response = self._request(
            "POST", headers={"Content-Type": "application/json"},
            json={"jsonrpc": "2.0", "method": method, "params": parameters, "id": str(uuid.uuid4())},
        )
        try:
            data = response.json()
        except ValueError as exc:
            raise ExampleError(f"{method} returned invalid JSON.") from exc
        if data.get("error"):
            raise ExampleError(f"{method} failed: {json.dumps(data['error'])}")
        return data.get("result")

    def search_jobs(self, networks: list[str], offset: int, page_size: int):
        result = self.call(
            "Scheduler.searchJobs",
            pageData={"offset": offset, "jobData": [], "pageSize": page_size, "total": 1},
            networks=networks, sortColumn="", descending=False,
        )
        if not isinstance(result, dict):
            raise ExampleError("Scheduler.searchJobs returned no page data.")
        return result

    def get_job(self, job_id: int):
        result = self.call("Scheduler.getJob", jobId=job_id)
        if not isinstance(result, dict):
            raise ExampleError(f"Scheduler.getJob returned no data for job ID {job_id}.")
        return result


def shallow_job_pages(client: Any, config: Config) -> Iterator[list[dict[str, Any]]]:
    offset = 0
    total: int | None = None
    while True:
        page = client.search_jobs(config.networks, offset, config.page_size)
        jobs = page.get("jobData") or []
        if not isinstance(jobs, list):
            raise ExampleError("Scheduler.searchJobs returned an invalid jobData collection.")
        yield jobs
        returned_page_size = int(page.get("pageSize") or config.page_size)
        if returned_page_size <= 0:
            raise ExampleError("Scheduler.searchJobs returned an invalid page size.")
        if total is None and page.get("total") is not None:
            total = int(page["total"])
        if total is not None and offset + len(jobs) >= total:
            return
        if total is None and len(jobs) < returned_page_size:
            return
        if not jobs:
            raise ExampleError("Scheduler.searchJobs returned an empty page before the reported total.")
        offset += returned_page_size


def collect_jobs(client: Any, config: Config) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    jobs: dict[int, dict[str, Any]] = {}
    failures: dict[int, dict[str, Any]] = {}
    for shallow_jobs in shallow_job_pages(client, config):
        for shallow in shallow_jobs:
            job_id = shallow.get("jobId")
            if not isinstance(job_id, int):
                raise ExampleError("Scheduler.searchJobs returned a job without an integer jobId.")
            if job_id in jobs or job_id in failures:
                continue
            try:
                jobs[job_id] = client.get_job(job_id)
            except Exception as exc:  # Preserve the remaining readable jobs.
                failures[job_id] = {
                    "jobId": job_id,
                    "jobName": shallow.get("jobName"),
                    "error": str(exc),
                }
    return [jobs[key] for key in sorted(jobs)], [failures[key] for key in sorted(failures)]


def utc_timestamp() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def build_documents(
    jobs: list[dict[str, Any]], failures: list[dict[str, Any]],
    networks: list[str], exported_at: str | None = None,
) -> tuple[dict[str, Any], dict[str, Any]]:
    timestamp = exported_at or utc_timestamp()
    backup = {
        "format": FORMAT_NAME,
        "formatVersion": FORMAT_VERSION,
        "exportedAt": timestamp,
        "networks": sorted(networks),
        "complete": not failures,
        "jobCount": len(jobs),
        "jobs": jobs,
    }
    failure_report = {
        "format": FAILURE_FORMAT_NAME,
        "formatVersion": FORMAT_VERSION,
        "exportedAt": timestamp,
        "failureCount": len(failures),
        "failures": failures,
    }
    return backup, failure_report


def write_json_atomic(path: Path, document: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary_path: Path | None = None
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", encoding="utf-8", dir=path.parent,
            prefix=f".{path.name}.", suffix=".tmp", delete=False,
        ) as output:
            temporary_path = Path(output.name)
            json.dump(document, output, indent=2, sort_keys=True, ensure_ascii=False)
            output.write("\n")
        os.replace(temporary_path, path)
    except Exception:
        if temporary_path is not None:
            temporary_path.unlink(missing_ok=True)
        raise


def backup_saved_jobs(client: Any, config: Config, exported_at: str | None = None) -> tuple[int, int]:
    jobs, failures = collect_jobs(client, config)
    backup, failure_report = build_documents(jobs, failures, config.networks, exported_at)
    write_json_atomic(config.output_path, backup)
    write_json_atomic(config.failure_path, failure_report)
    return len(jobs), len(failures)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--env", type=Path, default=Path(__file__).with_name(".env"))
    args = parser.parse_args()
    try:
        base = Path(__file__).resolve().parent
        config = Config.from_env(args.env, base)
        client = NetLDClient(config.base_url, config.api_key)
        client.login()
        jobs, failures = backup_saved_jobs(client, config)
        print(f"Wrote {jobs} complete saved jobs to {config.output_path}")
        print(f"Wrote {failures} retrieval failures to {config.failure_path}")
        return 2 if failures else 0
    except (ExampleError, OSError, ValueError, KeyError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: Examples/Advanced/Backup-Saved-Jobs/nodeJS/backup-saved-jobs.mjs
#!/usr/bin/env node

import { randomUUID } from "node:crypto";
import { mkdir, readFile, rename, rm, writeFile } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { fileURLToPath, pathToFileURL } from "node:url";

export const FORMAT_NAME = "logicvein-netld-saved-job-backup";
export const FAILURE_FORMAT_NAME = "logicvein-netld-saved-job-backup-failures";
export const FORMAT_VERSION = 1;

export class ExampleError extends Error {}

export function parseNetworks(value) {
  const networks = [...new Set(value.split(",").map((item) => item.trim()).filter(Boolean))].sort();
  if (!networks.length) throw new ExampleError("NETLD_NETWORKS must contain at least one managed network.");
  return networks;
}

async function loadDotEnv(file) {
  let text;
  try { text = await readFile(file, "utf8"); }
  catch (error) { if (error.code === "ENOENT") return; throw error; }
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#") || !line.includes("=")) continue;
    const offset = line.indexOf("=");
    process.env[line.slice(0, offset).trim()] = line.slice(offset + 1).trim().replace(/^(['"])(.*)\1$/, "$2");
  }
}

async function loadConfig(directory, envFile) {
  await loadDotEnv(envFile);
  for (const name of ["NETLD_BASE_URL", "NETLD_API_KEY"]) {
    if (!process.env[name]?.trim()) throw new ExampleError(`Set ${name} in the environment file.`);
  }
  const pageSize = Number.parseInt(process.env.NETLD_JOB_PAGE_SIZE || "100", 10);
  if (!Number.isInteger(pageSize) || pageSize <= 0) throw new ExampleError("NETLD_JOB_PAGE_SIZE must be a positive integer.");
  return {
    baseUrl: process.env.NETLD_BASE_URL.replace(/\/$/, ""), apiKey: process.env.NETLD_API_KEY,
    networks: parseNetworks(process.env.NETLD_NETWORKS || "Default"), pageSize,
    outputFile: path.resolve(directory, process.env.NETLD_OUTPUT_FILE || "saved-jobs.json"),
    failureFile: path.resolve(directory, process.env.NETLD_FAILURE_FILE || "saved-job-failures.json"),
  };
}

class NetLDClient {
  constructor(baseUrl, apiKey, timeout = 30000) {
    this.baseUrl = baseUrl; this.apiKey = apiKey; this.timeout = timeout; this.cookies = "";
  }
  headers(extra = {}) {
    return { Authorization: `Bearer ${this.apiKey}`, ...(this.cookies ? { Cookie: this.cookies } : {}), ...extra };
  }
  async request(options = {}) {
    let response;
    try {
      response = await fetch(`${this.baseUrl}/rest`, {
        redirect: "manual", signal: AbortSignal.timeout(this.timeout), ...options,
        headers: this.headers(options.headers),
      });
    } catch { throw new ExampleError(`Could not reach ${this.baseUrl}.`); }
    if (response.status >= 300 && response.status < 400) {
      throw new ExampleError(`Request redirected to ${response.headers.get("location") || ""}.`);
    }
    if (!response.ok) throw new ExampleError(`Request failed with HTTP ${response.status}.`);
    const setCookies = response.headers.getSetCookie?.() || splitSetCookie(response.headers.get("set-cookie") || "");
    if (setCookies.length) this.cookies = setCookies.map((item) => item.split(";", 1)[0]).join("; ");
    return response;
  }
  async login() { await this.request(); }
  async call(method, parameters = {}) {
    const response = await this.request({
      method: "POST", headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ jsonrpc: "2.0", method, params: parameters, id: randomUUID() }),
    });
    const data = await response.json();
    if (data.error) throw new ExampleError(`${method} failed: ${JSON.stringify(data.error)}`);
    return data.result;
  }
  async searchJobs(networks, offset, pageSize) {
    const result = await this.call("Scheduler.searchJobs", {
      pageData: { offset, jobData: [], pageSize, total: 1 },
      networks, sortColumn: "", descending: false,
    });
    if (!result || typeof result !== "object") throw new ExampleError("Scheduler.searchJobs returned no page data.");
    return result;
  }
  async getJob(jobId) {
    const result = await this.call("Scheduler.getJob", { jobId });
    if (!result || typeof result !== "object") throw new ExampleError(`Scheduler.getJob returned no data for job ID ${jobId}.`);
    return result;
  }
}

function splitSetCookie(header) { return header ? header.split(/,(?=\s*[^;=]+=[^;]+)/) : []; }

export async function collectJobs(client, config) {
  const jobs = new Map();
  const failures = new Map();
  let offset = 0;
  let total;
  while (true) {
    const page = await client.searchJobs(config.networks, offset, config.pageSize);
    const shallowJobs = page.jobData || [];
    if (!Array.isArray(shallowJobs)) throw new ExampleError("Scheduler.searchJobs returned an invalid jobData collection.");
    for (const shallow of shallowJobs) {
      const jobId = shallow.jobId;
      if (!Number.isInteger(jobId)) throw new ExampleError("Scheduler.searchJobs returned a job without an integer jobId.");
      if (jobs.has(jobId) || failures.has(jobId)) continue;
      try { jobs.set(jobId, await client.getJob(jobId)); }
      catch (error) { failures.set(jobId, { jobId, jobName: shallow.jobName, error: error.message }); }
    }
    const returnedPageSize = Number(page.pageSize || config.pageSize);
    if (!Number.isInteger(returnedPageSize) || returnedPageSize <= 0) throw new ExampleError("Scheduler.searchJobs returned an invalid page size.");
    if (total == null && page.total != null) total = Number(page.total);
    if (total != null && offset + shallowJobs.length >= total) break;
    if (total == null && shallowJobs.length < returnedPageSize) break;
    if (!shallowJobs.length) throw new ExampleError("Scheduler.searchJobs returned an empty page before the reported total.");
    offset += returnedPageSize;
  }
  return {
    jobs: [...jobs.entries()].sort(([a], [b]) => a - b).map(([, job]) => job),
    failures: [...failures.entries()].sort(([a], [b]) => a - b).map(([, failure]) => failure),
  };
}

export function buildDocuments(jobs, failures, networks, exportedAt = new Date().toISOString().replace(/\.\d{3}Z$/, "Z")) {
  return {
    backup: {
      format: FORMAT_NAME, formatVersion: FORMAT_VERSION, exportedAt,
      networks: [...networks].sort(), complete: failures.length === 0,
      jobCount: jobs.length, jobs,
    },
    failureReport: {
      format: FAILURE_FORMAT_NAME, formatVersion: FORMAT_VERSION, exportedAt,
      failureCount: failures.length, failures,
    },
  };
}

function canonicalize(value) {
  if (Array.isArray(value)) return value.map(canonicalize);
  if (value && typeof value === "object") {
    return Object.fromEntries(Object.keys(value).sort().map((key) => [key, canonicalize(value[key])]));
  }
  return value;
}

async function writeJsonAtomic(destination, document) {
  await mkdir(path.dirname(destination), { recursive: true });
  const temporary = path.join(path.dirname(destination), `.${path.basename(destination)}.${randomUUID()}.tmp`);
  try {
    await writeFile(temporary, `${JSON.stringify(canonicalize(document), null, 2)}\n`, {
      encoding: "utf8", flag: "wx", mode: 0o600,
    });
    await rename(temporary, destination);
  } catch (error) {
    await rm(temporary, { force: true }).catch(() => {});
    throw error;
  }
}

export async function backupSavedJobs(client, config, exportedAt) {
  const { jobs, failures } = await collectJobs(client, config);
  const { backup, failureReport } = buildDocuments(jobs, failures, config.networks, exportedAt);
  await writeJsonAtomic(config.outputFile, backup);
  await writeJsonAtomic(config.failureFile, failureReport);
  return { jobCount: jobs.length, failureCount: failures.length };
}

export async function main() {
  const directory = path.dirname(fileURLToPath(import.meta.url));
  const envOffset = process.argv.indexOf("--env");
  if (envOffset >= 0 && !process.argv[envOffset + 1]) throw new ExampleError("--env requires a file path.");
  const envFile = envOffset >= 0 ? path.resolve(process.argv[envOffset + 1]) : path.join(directory, ".env");
  const config = await loadConfig(directory, envFile);
  const client = new NetLDClient(config.baseUrl, config.apiKey);
  await client.login();
  const result = await backupSavedJobs(client, config);
  console.log(`Wrote ${result.jobCount} complete saved jobs to ${config.outputFile}`);
  console.log(`Wrote ${result.failureCount} retrieval failures to ${config.failureFile}`);
  return result.failureCount ? 2 : 0;
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().then((code) => { process.exitCode = code; }).catch((error) => {
    console.error(`Error: ${error.message}`); process.exitCode = 1;
  });
}
